File: c++/2_Add_Two_Numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *dummy = new ListNode();
        ListNode *curr = dummy;
        int carry = 0;
        int val;
        while (l1 && l2) {
            val = (l1->val + l2->val + carry) % 10;
            carry = (l1->val + l2->val + carry) / 10;
            curr->next = new ListNode(val=val);
            curr = curr->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1) {
            val = (l1->val + carry) % 10;
            carry = (l1->val + carry) / 10;
            curr->next = new ListNode(val=val);
            curr = curr->next;
            l1 = l1->next;
        }
        while (l2) {
            val = (l2->val + carry) % 10;
            carry = (l2->val + carry) / 10;
            curr->next = new ListNode(val=val);
            curr = curr->next;
            l2 = l2->next;
        }
        if (carry)
            curr->next = new ListNode(val=carry);
        return dummy->next;
    }
};

File: rust/2_Add_Two_Numbers/solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut l1 = l1;
        let mut l2 = l2;
        let mut dummy = Some(Box::new(ListNode::new(-1)));
        let mut curr = dummy.as_mut().unwrap();
        let mut carry = 0;
        while l1 != None && l2 != None {
            let val = (l1.as_ref().unwrap().val + l2.as_ref().unwrap().val + carry) % 10;
            carry = (l1.as_ref().unwrap().val + l2.as_ref().unwrap().val + carry) / 10;
            curr.next = Some(Box::new(ListNode::new(val)));
            l1 = l1.unwrap().next;
            l2 = l2.unwrap().next;
            curr = curr.next.as_mut().unwrap();
        }
        while l1 != None {
            let val = (l1.as_ref().unwrap().val + carry) % 10;
            carry = (l1.as_ref().unwrap().val + carry) / 10;
            curr.next = Some(Box::new(ListNode::new(val)));
            l1 = l1.unwrap().next;
            curr = curr.next.as_mut().unwrap();
        }
        while l2 != None {
            let val = (l2.as_ref().unwrap().val + carry) % 10;
            carry = (l2.as_ref().unwrap().val + carry) / 10;
            curr.next = Some(Box::new(ListNode::new(val)));
            l2 = l2.unwrap().next;
            curr = curr.next.as_mut().unwrap();
        }
        if carry != 0 {
            curr.next = Some(Box::new(ListNode::new(carry)));
        }
        dummy.unwrap().next
    }
}
